elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+ * <p>
+ * {@link SplitFilterExpressionVisitor} is leveraged by the AggregationStore to construct the JPQL query.
+ * {@link FilterExpression} for AggregationDataStore must be split into those that apply to metric aggregations
+ * ({@code HAVING} clauses) and those that apply to dimensions ({@code WHERE} clauses), although only a single
+ * {@link FilterExpression} is passed to the datastore in each query. The split groups {@code HAVING} clauses and
+ * {@code WHERE} clauses separately. For example:
+ * <pre>
+ * +-------------------+-----------------------------+------------------------+
+ * |    Expression     |             SQL             | WHERE-clause Promotion |
+ * +-------------------+-----------------------------+------------------------+
+ * | H1 AND W1 AND H2  | WHERE W1 HAVING (H1 AND H2) | No                     |
+ * | (H1 OR H2) AND W1 | WHERE W1 HAVING (H1 OR H2)  | No                     |
+ * | H1 OR W1          | HAVING (H1 OR W1)           | Yes                    |
+ * | (W1 AND H1) OR W2 | HAVING ((W1 AND H1) OR W2)  | Yes                    |
+ * +-------------------+-----------------------------+------------------------+
+ * </pre>
+ * Note that {@link SplitFilterExpressionVisitor} might incur more-than-expected network I/O in the case of WHERE-clause
+ * promotion.
+ * <p>
+ * {@link SplitFilterExpressionVisitor} splits by storing {@code WHERE} and {@code HAVING} clauses in
+ * {@link Pair#getLeft() left} and {@link Pair#getRight() right} of a {@link Pair}, respectively. For example:
+ * <pre>
+ * {@code
+ * Pair<FilterExpression, FilterExpression> filterPair = filterExpression.accept(splitFilterExpressionVisitor);
+ *
+ * FilterExpression whereClauseFilter = filterPair.getLeft();
+ * FilterExpression havingClauseFilter = filterPair.getRight();
+ * }
+ * </pre>
+ * {@link SplitFilterExpressionVisitor} is thread-safe and can be accessed by multiple threads at the same time.
+ */
+@Slf4j
+public class SplitFilterExpressionVisitor implements FilterExpressionVisitor<FilterConstraints> {
+
+    @Getter(value = AccessLevel.PRIVATE)
+    private final Schema schema;
+    @Getter(value = AccessLevel.PRIVATE)
+    private final FilterExpressionNormalizationVisitor normalizationVisitor;
+
+    /**
+     * Constructor.
+     *
+     * @param schema  Object that offers meta information about an entity field
+     * @param normalizationVisitor  A helper {@link FilterExpressionVisitor} that is leveraged to transform
+     * expressions like {@code NOT (A AND B)} into {@code NOT A OR NOT B}
+     *
+     * @throws NullPointerException if any one of the argument is {@code null}
+     */
+    public SplitFilterExpressionVisitor(
+            final Schema schema,
+            final FilterExpressionNormalizationVisitor normalizationVisitor
+    ) {
+        this.schema = Objects.requireNonNull(schema, "schema");
+        this.normalizationVisitor = Objects.requireNonNull(normalizationVisitor, "normalizationVisitor");
```

Comment:
This could probably just be constructed in the constructor rather than passed in as an argument.

Reply:
Yep.

How about putting it as a static field?